noir-projects/aztec-nr/aztec/src/note/utils.nr

```diff
@@ -29,66 +19,55 @@ fn compute_inner_note_hash<Note, N, M>(note: Note) -> Field where Note: NoteInte
     context: &mut PrivateContext
 ) -> Field where Note: NoteInterface<N, M> {
     let header = note_with_header.get_header();
     let (_, inner_nullifier) = note_with_header.compute_note_hash_and_nullifier(context);
 
-    let input = [header.contract_address.to_field(), inner_nullifier];
-    pedersen_hash(input, GENERATOR_INDEX__OUTER_NULLIFIER)
+    compute_siloed_nullifier_from_preimage(header.contract_address, inner_nullifier)
 }
 
-pub fn compute_note_hash_for_insertion<Note, N, M>(note: Note) -> Field where Note: NoteInterface<N, M> {
-    compute_inner_note_hash(note)
-}
-
-pub fn compute_note_hash_for_read_request<Note, N, M>(note: Note) -> Field where Note: NoteInterface<N, M> {
-    let header = note.get_header();
-
-    let inner_note_hash = compute_inner_note_hash(note);
-
+fn compute_note_hash_for_read_request_from_innter_and_nonce(
+    inner_note_hash: Field,
+    nonce: Field
+) -> Field {
     // TODO(#1386): This if-else can be nuked once we have nonces injected from public
-    if (header.nonce == 0) {
+    if (nonce == 0) {
         // If nonce is zero, that means we are reading a public note.
         inner_note_hash
     } else {
-        compute_unique_hash(header.nonce, inner_note_hash)
+        compute_unique_note_hash(nonce, inner_note_hash)
     }
 }
 
+pub fn compute_note_hash_for_read_request<Note, N, M>(note: Note) -> Field where Note: NoteInterface<N, M> {
+    let inner_note_hash = compute_inner_note_hash(note);
+    let nonce = note.get_header().nonce;
+
+    compute_note_hash_for_read_request_from_innter_and_nonce(inner_note_hash, nonce)
+}
+
 pub fn compute_note_hash_for_consumption<Note, N, M>(note: Note) -> Field where Note: NoteInterface<N, M> {
     let header = note.get_header();
     // There are 3 cases for reading a note intended for consumption:
     // 1. The note was inserted in this transaction, and is transient.
     // 2. The note was inserted in a previous transaction, and was inserted in public
     // 3. The note was inserted in a previous transaction, and was inserted in private
 
+    let inner_note_hash = compute_inner_note_hash(note);
+
     if (header.note_hash_counter != 0) {
         // If a note is transient, we just read the inner_note_hash (kernel will silo by contract address).
-        compute_inner_note_hash(note)
+        inner_note_hash
     } else {
         // If a note is not transient, that means we are reading a settled note (from tree) created in a
         // previous TX. So we need the siloed_note_hash which has already been hashed with
         // nonce and then contract address. This hash will match the existing leaf in the note hash
         // tree, so the kernel can just perform a membership check directly on this hash/leaf.
-        compute_siloed_note_hash(note)
+        let unique_note_hash = compute_note_hash_for_read_request_from_innter_and_nonce(inner_note_hash, header.nonce);
+        compute_siloed_note_hash(header.contract_address, unique_note_hash)
```

Comment:
This looks uglier but it was necessary to be able to compute the inner note hash only once.